addons/binding/org.openhab.binding.elerotransmitterstick/src/main/java/org/openhab/binding/elerotransmitterstick/internal/EleroTransmitterStickHandlerFactory.java

```diff
@@ -0,0 +1,78 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.elerotransmitterstick.internal;
+
+import static org.openhab.binding.elerotransmitterstick.EleroTransmitterStickBindingConstants.*;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Hashtable;
+import java.util.Map;
+import java.util.Set;
+
+import org.eclipse.smarthome.config.discovery.DiscoveryService;
+import org.eclipse.smarthome.core.thing.Bridge;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandlerFactory;
+import org.eclipse.smarthome.core.thing.binding.ThingHandler;
+import org.openhab.binding.elerotransmitterstick.discovery.EleroChannelDiscoveryService;
+import org.openhab.binding.elerotransmitterstick.handler.EleroChannelHandler;
+import org.openhab.binding.elerotransmitterstick.handler.EleroGroupHandler;
+import org.openhab.binding.elerotransmitterstick.handler.EleroTransmitterStickHandler;
+import org.osgi.framework.ServiceRegistration;
+
+/**
+ * The {@link EleroTransmitterStickHandlerFactory} is responsible for creating things and thing
+ * handlers.
+ *
+ * @author Volker Bier - Initial contribution
+ */
+public class EleroTransmitterStickHandlerFactory extends BaseThingHandlerFactory {
+
+    private final static Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = new HashSet<>(2);
+    static {
+        SUPPORTED_THING_TYPES_UIDS.add(THING_TYPE_STICK);
+        SUPPORTED_THING_TYPES_UIDS.add(THING_TYPE_ELERO_CHANNEL);
+        SUPPORTED_THING_TYPES_UIDS.add(THING_TYPE_ELERO_GROUP);
+    }
+
+    private Map<ThingUID, ServiceRegistration<?>> discoveryServiceRegs = new HashMap<>();
+
+    @Override
+    public boolean supportsThingType(ThingTypeUID thingTypeUID) {
+        return SUPPORTED_THING_TYPES_UIDS.contains(thingTypeUID);
+    }
+
+    @Override
+    protected ThingHandler createHandler(Thing thing) {
+        ThingTypeUID thingTypeUID = thing.getThingTypeUID();
+
+        if (thingTypeUID.equals(THING_TYPE_STICK)) {
+            EleroTransmitterStickHandler bridgeHandler = new EleroTransmitterStickHandler((Bridge) thing);
+
+            EleroChannelDiscoveryService discoveryService = new EleroChannelDiscoveryService(bridgeHandler);
```

Comment:
The discovery service is an independent OSGi service which acts totally independent of this handler. Please remove this dependency.

Reply:
AFAICS, this is how all other plugins do it. I have no idea how to do this differently.

Reply:
Which other plugin is creating a discoveryservice inside the `createHandler` method? If such a binding exists, this is clearly a bug and and issue should be created to fix it. Have a look at the [discovery service documentation](https://www.eclipse.org/smarthome/documentation/development/bindings/discovery-services.html) on how to implement a discovery service. Here you do not need to create that instance.

Reply:
lots of openhab bindings do it this way: avmfritz, dscalarm, freebox, ...

Reply:
Hmm "Bindings can implement the DiscoveryService interface and register it as an OSGi service" indeed for people who are not familiar with OSGi need an example here... Meanwhile you can have a look at the [astro binding](https://github.com/eclipse/smarthome/blob/master/extensions/binding/org.eclipse.smarthome.binding.astro/src/main/java/org/eclipse/smarthome/binding/astro/internal/discovery/AstroDiscoveryService.java#L39). Just creating this annotation already enables it to be a registered OSGi service.

Inside the Handler you do not need to start it as it will be started once the user had added your bridge thing and clicks on "scan" in the INBOX in the paperUI for example.

Can you please point me to the exact file and line where the avmfritz binding is doing that? I cannot find it.

Reply:
See https://github.com/openhab/openhab2-addons/blob/master/addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/HandlerFactory.java#L70

Reply:
There is no need to pass around the `bridgeHandler` to the discovery service. The discovery service only creates `DiscoveryResult`s which are totally independent of concrete `thing`s. Maybe the [hue binding](https://github.com/eclipse/smarthome/blob/master/extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/discovery/HueLightDiscoveryService.java) is a good example since it uses a bridge too and has a discovery service for its connected lights.